### PR TITLE
feat: add scenario-based open-loop validation metrics and integration

### PR DESCRIPTION
**Pull request by TatsukiNishimura**
This pull request introduces a modular scenario-based open-loop validation system to the `diffusion_planner` codebase.

  The system evaluates model predictions against custom NPZ datasets using metric-specific configurations. It currently supports centerline and departure metrics, integrates with both training and standalone validation workflows, and produces
  aggregate summaries, per-sample details, and optional visualizations.

  The terminology has been updated from “Override Open-loop” to “Scenario-based Open-loop” because the evaluation is intended for scenario-specific validation rather than model overrides.

  ## Scenario-based Open-loop Validation

  * Added the `scenario_based_open_loop` module, including:
    * NPZ list loading and validation
    * Metric registry
    * Open-loop validation execution
    * Training-loop integration
    * Prediction visualization
  * Added centerline metrics in `planner_metrics/centerline.py`:
    * Centerline ADE
    * Centerline FDE
    * Batched nearest-distance computation
  * Added departure metrics in `planner_metrics/departure.py`:
    * Departure failure rate
    * Maximum displacement
    * Per-sample departure diagnostics
  * Added the shared `MetricEvaluation` result type in `planner_metrics/evaluation.py`, allowing metric implementations to return both aggregateable scores and metric-specific details without coupling the runner to individual metrics.

  ## Training and Validation Integration

  * Added scenario-based open-loop configuration fields to `TrainConfig`:
    * `scenario_based_open_loop_list`
    * `scenario_centerline_horizon_seconds`
    * `scenario_departure_horizon_seconds`
    * `scenario_departure_minimum_displacement_m`
  * Added corresponding command-line arguments to `train_predictor.py`, `train_run.py`, `valid_predictor.py`, and `valid_run.py`.
  * Integrated scenario-based open-loop validation into the training loop.
  * Added standalone validation support with:
    * `--scenario_based_open_loop_list`
    * `--scenario_based_open_loop_only`
  * Summaries are logged to W&B during training and written to `summary.json` during standalone validation.
  * Optional per-sample details are written as JSONL files.
  * Optional scene visualizations are generated for each evaluated sample.

  ## Configuration and Input Validation

  * Added validation for the metric-to-NPZ JSON mapping:
    * JSON object structure
    * Supported metric names
    * List-of-string values
    * NPZ file existence
    * NPZ readability
  * The training launcher validates the scenario-based open-loop list before starting the training process.
  * `DiffusionPlannerData` now accepts both the existing JSON list path format and an in-memory list of NPZ paths, allowing the open-loop evaluator to reuse the shared dataset implementation.

  ## Metric Extensibility

  * The validation runner is metric-agnostic.
  * Metric-specific computation and diagnostic details are owned by the scorer implementation.
  * Adding a new metric does not require adding metric-specific branches to the validation runner.

  ## Testing

  Added tests for:

  * Centerline projection, ADE, and FDE
  * Departure failure-rate calculation
  * Scenario-based open-loop JSON and NPZ validation
  * Unsupported metric rejection
  * Metric parameter extraction
  * Metric registry contents

  The relevant tests pass successfully.

### CODE REVIEW

**Comment by go-sakayori** [comment]
@TatsukiNishimura 
you've implemented the metrics below `override_validation/metrics`, I think it is better to implement in `planner_metrics`. The metrcis are not only for override scenes, and it can be used to any scenario. There are many duplication so please consider using the existing functions
- _point_to_polylines_min_dist in centerline.py is essentially the same as planner_metrics/geometry.py::_point_to_segments_min_dist (which also has OOM-safe chunking)
- planner_metrics/subscores.py already has compute_centerline_score_batch and compute_lane_departure_penalty following the (ego_trajs, data, ...) -> (N,) convention

File: diffusion_planner/train_run.py [hunk 1]
```diff
@@ -32,6 +33,16 @@ def parse_args() -> argparse.Namespace:
         help="optional: dir tree of route NPZ frames for closed-loop validation, OR a .json path "
         "list of such dirs (like --train_set_list). Empty = disabled.",
     )
+    p.add_argument(
+        "--override_open_loop_list",
+        default="",
+        help="optional JSON mapping override Open-loop metric names to NPZ path lists. Empty = disabled.",
+    )
+    p.add_argument(
+        "--override_open_loop_config",
```
Comment (go-sakayori) [inline diffusion_planner/train_run.py]: Why not just implement in `train_config.py`.

File: diffusion_planner/diffusion_planner/override_validation/open_loop.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        )
+    if not list_path:
+        return {}, {}
+
+    raw_lists = _load_json_object(list_path, "list")
+    raw_config = _load_json_object(config_path, "config")
+    metric_parameters = raw_config.get("metrics", {})
+    if not isinstance(metric_parameters, dict):
+        raise ValueError("Override Open-loop config field 'metrics' must be an object")
+    interval = raw_config.get("interval_epochs", 1)
```
Comment (go-sakayori) [inline diffusion_planner/diffusion_planner/override_validation/open_loop.py]: We are going to evaluate each epoch, so I don't think we need this field

File: planner_metrics/geometry.py [hunk 3]
```diff
@@ -723,6 +723,20 @@ def _point_to_segments_min_dist(
     return torch.cat(results)
 
 
+def _point_to_polylines_min_dist(
```
Comment (go-sakayori) [inline planner_metrics/geometry.py]: I think it's a dead code, and only used in test. 
is it used anywhere else?

File: diffusion_planner/train_run.py [hunk 4]
```diff
@@ -32,6 +32,11 @@ def parse_args() -> argparse.Namespace:
         help="optional: dir tree of route NPZ frames for closed-loop validation, OR a .json path "
         "list of such dirs (like --train_set_list). Empty = disabled.",
     )
+    p.add_argument(
+        "--override_open_loop_list",
```
Comment (go-sakayori) [inline diffusion_planner/train_run.py]: It might not be override specific. I would rather use other names. ex: `scenario_base_open_loop`

File: diffusion_planner/train_run.py [hunk 5]
```diff
@@ -42,6 +47,11 @@ def main() -> None:
     save_path = Path(args.output_root) / f"{datetime.now():%Y%m%d-%H%M%S}_{args.exp_name}"
     save_path.mkdir(parents=True, exist_ok=True)
 
+    if args.override_open_loop_list:
+        override_list = Path(args.override_open_loop_list).resolve()
+        if not override_list.is_file():
+            raise FileNotFoundError(f"Override Open-loop list not found: {override_list}")
```
Comment (go-sakayori) [inline diffusion_planner/train_run.py]: This only checks if the file exists or not. Shouldn't we check the content of the file before training.

File: diffusion_planner/diffusion_planner/override_validation/validate.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+from diffusion_planner.override_validation.open_loop import run_override_open_loop_validation
+from diffusion_planner.utils import ddp
+
+
+def override_open_loop_validate(model, args, epoch: int) -> None:
+    """Run configured Override Open-loop metrics and publish summaries to W&B.
+
+    Called by rank 0 only.  The list JSON selects NPZs per metric, while the
+    companion config determines the validation cadence and metric parameters.
```
Comment (go-sakayori) [inline diffusion_planner/diffusion_planner/override_validation/validate.py]: I think this docstring is old. Also please update the PR description `Summaries are logged to W&B at configurable intervals`.

File: diffusion_planner/diffusion_planner/override_validation/open_loop.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        prefix = f"override_{metric_name}_"
+        parameters[metric_name] = {
+            field_name[len(prefix) :]: value
+            for field_name, value in values.items()
+            if field_name.startswith(prefix)
+        }
+    return parameters
+
+
+class _NpzPathDataset(Dataset):
```
Comment (go-sakayori) [inline diffusion_planner/diffusion_planner/override_validation/open_loop.py]: Same function as `DiffusionPlannerData.__getitem__ `. Considering removing duplicated code

File: diffusion_planner/diffusion_planner/override_validation/open_loop.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+                # physical coordinates and metrics receive denormalized inputs.
+                metric_inputs = args.observation_normalizer.inverse(prepared.inputs)
+                batch_size = int(outputs["prediction"].shape[0])
+                ego_prediction = outputs["prediction"][:, 0]
+                batch_start = count
+                count += batch_size
+                per_sample_scores = scorer(ego_prediction, metric_inputs, parameters)
+                departure_displacements = (
+                    departure_max_displacement(ego_prediction, metric_inputs, parameters)
+                    if metric_name == "departure"
```
Comment (go-sakayori) [inline diffusion_planner/diffusion_planner/override_validation/open_loop.py]: The docstring in open_loop.py states that "the scorer registry owns metric-specific computation", but the implementation actually contains if metric_name == "departure" branches in two places here and the detail-dict construction around line 180). This means every new metric will require modifying the runner itself, which defeats the purpose of the registry. I recommend moving this responsibility to the scorer side .

**Review by go-sakayori — CHANGES_REQUESTED** [review]
Please update the PR description when you modify the code

File: diffusion_planner/diffusion_planner/override_validation/open_loop.py (hunk 7 above)
Comment (TatsukiNishimura) [reply diffusion_planner/diffusion_planner/override_validation/open_loop.py]: ~Thanks for pointing this out. `_NpzPathDataset `intentionally accepts an in-memory list[str] of NPZ paths for each metric, whereas `DiffusionPlannerData` expects a JSON file path and loads the list itself. Reusing it directly would require writing temporary JSON files or changing `DiffusionPlannerData`,
  so I’ll keep the small adapter here to avoid modifying the shared dataset implementation.~

 Thanks for pointing this out. I’ve updated DiffusionPlannerData to accept both the existing JSON list path and an in-memory list[str] of NPZ paths. Open-loop validation now reuses DiffusionPlannerData directly, so the duplicated _NpzPathDataset adapter has been removed.

File: planner_metrics/departure.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    horizon_steps: int | None = None,
+) -> torch.Tensor:
+    """Return 1.0 for failure and 0.0 for successful departure, shape ``(N,)``."""
+    if minimum_displacement_m < 0:
+        raise ValueError("minimum_displacement_m must be non-negative")
+    maximum = compute_max_displacement_batch(ego_trajs, data, horizon_steps)
+    return (maximum < minimum_displacement_m).to(ego_trajs.dtype)
+
+
+def departure_max_displacement(
```
Comment (go-sakayori) [inline planner_metrics/departure.py]: Is this going to be used in the near future? Not currently used.

File: diffusion_planner/valid_run.py [hunk 10]
```diff
@@ -48,10 +79,23 @@ def main() -> None:
         str(model_path),
         "--args_json_path",
         str(args_json_path),
+        "--batch_size",
+        str(args.batch_size),
         "--save_predictions_dir",
         str(save_dir),
     ]
-    rc = tee_run(cmd, cwd=here, log_path=save_dir.parent / "valid_log.txt")
+    if args.valid_set_list:
```
Comment (go-sakayori) [inline diffusion_planner/valid_run.py:84]: The base cmd list above still passes `--valid_set_list` unconditionally (L76-77), so this conditional adds the flag a second time.
- When `--valid_set_list` is given, the flag is passed twice. Harmless only because argparse keeps the last value.
- When it is omitted with `--scenario_based_open_loop_only`, the leftover pair passes an empty string, so valid_predictor.py receives valid_set_list="" instead of None. This silently bypasses the if valid_cfg.valid_set_list is None guard in run_validation; today the scenario-only early return happens to hide it, but any future reordering would surface a confusing DiffusionPlannerData("") failure instead of the intended error message.
 
Consider removing the unconditional "--valid_set_list", valid_set_list pair from the base list (the valid_set_list = args.valid_set_list assignment on L48 also becomes unused) and keep only this conditional.

**Review by go-sakayori — APPROVED** [review]
Please update the PR title.
